ml/dgl_utils.py

```diff
@@ -0,0 +1,253 @@
+        self._check_nodes_lineup_with_edges()
+
+    def _check_nodes_lineup_with_edges(self):
+        nodes = self.ndf[self.node]
+        unique_nodes = nodes.unique()
+        logger.info(
+            f"{len(nodes)} entities from column {self.node}\n with {len(unique_nodes)} unique entities"
+        )
+        if len(nodes) != len(unique_nodes):  # why would this be so?
+            logger.warning(
+                f"Nodes DataFrame has duplicate entries for column {self.node}"
+            )
+            # logger.warning(f'** Dropping duplicates in {self.node} and reassigning to self.ndf')
+            # self.ndf = self.ndf.drop_duplicates(subset=[self.node])  # this might do damage to the representation
+        # now check that self.entity_to_index is in 1-1 to with self.ndf[self.node]
+        nodes = self.ndf[self.node]
+        res = nodes.isin(self.entity_to_index)
+        if res.sum() != len(nodes):
+            logger.warning(
+                f"Some Edges connect to Nodes not explicitly mentioned in nodes DataFrame (ndf)"
+            )
+        if len(self.entity_to_index) > len(nodes):
+            logger.warning(
+                f"There are more entities in edges DataFrame (edf) than in nodes DataFrame (ndf)"
+            )
+
+    def _node_featurization(self):
+        logger.info("Running Node Featurization")
+        ndf = self.ndf
+        if not self.use_node_label_as_feature:
+            logger.info(
+                f"Dropping {self.node} from DataFrame so not to include it as an explicit feature, "
+                f"which would just add np.eye({len(ndf), len(ndf)}) to the feature matrix under sklearn,"
+                f"which might be useful for factorization machines"
+                f"or under dirty_cat would mix rows with similar names, which seems odd"
+            )
+            ndf = ndf.drop(columns=[self.node])
+        ndata, node_encoders = featurization(ndf, self.node_target, self.vectorizer)
+        self.ndata = ndata
+        self.node_encoders = node_encoders
+
+    def _edge_featurization(self):
+        logger.info("Running Edge Featurization")
+        edata, edge_encoders = featurization(
+            self.edf, self.edge_target, self.vectorizer
+        )
+        self.edata = edata
+        self.edge_encoders = edge_encoders
+
+    def build_simple_graph(self):
+        self._convert_edgeDF_to_DGL()
+
+    def embeddings(self):
+        # here we make node and edge features and add them to the DGL graph instance self.graph
+        if not hasattr(self, "graph"):
+            self._convert_edgeDF_to_DGL()
+        if not hasattr(self, "ndata"):
+            self._node_featurization()
+            # first add the ndata
+            self.graph.ndata.update(self.ndata)
+        if not hasattr(self, "edata"):
+            self._edge_featurization()
+            # then add any edata
+            self.graph.edata.update(self.edata)
+
+    def umap(self, scale=False, show=False, color_labels=None):
```

Comment:
this seems tied to DGL... but needn't be, right? `g.nodes(...).umap(engine='umap_learn').plot()`

Reply:
our umap signature will be interesting!

-- passthrough args to engine (esp umap_learn vs cuml.umap)
-- smart defaults
-- info for cleaning vs settings vs optional training vs plotting...